### PR TITLE
Settings data defined

### PR DESCRIPTION
**Pull request by zakalawe**
This won’t build yet, since I didn’t include commented-out code
to compile the rest of the plugin.

### CODE REVIEW

File: source/AsyncRgbLedAnalyzerSettings.cpp [hunk 1]
```diff
@@ -67,7 +75,33 @@ const char* AsyncRgbLedAnalyzerSettings::SaveSettings()
...
-	text_archive << mBitRate;
+	text_archive << mLEDController;
 
 	return SetReturnString( text_archive.GetString() );
 }
+
+
+U8 AsyncRgbLedAnalyzerSettings::BitSize() const
+{
+	if (mLEDController == LED_LPD1886_12bit)
```
Comment (Marcus10110) [inline source/AsyncRgbLedAnalyzerSettings.cpp]: the pattern here should probably match with the pattern in LEDChannelCount, either using a switch statement or an if statement.

File: source/AsyncRgbLedAnalyzerSettings.h [hunk 2]
```diff
@@ -10,18 +10,33 @@ class AsyncRgbLedAnalyzerSettings : public AnalyzerSettings
...
+
+	/// bits ber LED channel, either 8 or 10 at present
+	U8 BitSize() const; 
+
+	/// LED channel count, either 3 (RGB) or 9 (three RGB outputs) at present
+	U8 LEDChannelCount() const;
 protected:
 	std::auto_ptr< AnalyzerSettingInterfaceChannel >	mInputChannelInterface;
-	std::auto_ptr< AnalyzerSettingInterfaceInteger >	mBitRateInterface;
+	std::auto_ptr< AnalyzerSettingInterfaceNumberList >	mControllerInterface;
```
Comment (Marcus10110) [inline source/AsyncRgbLedAnalyzerSettings.h]: Lets change this to std::unique_ptr.
I'd like to stop using std::auto_ptr. I think we put that in there very early on (over 8 years ago) when we were removing the boost dependency from the analyzers and wanted a replacement for boost::scoped_ptr.
auto_ptr probably never made sense here. 
I'll check in on our build system support for this, and I'll have more notes in my overall review notes.

File: source/AsyncRgbLedAnalyzerSettings.h [hunk 3]
```diff
@@ -10,18 +10,33 @@ class AsyncRgbLedAnalyzerSettings : public AnalyzerSettings
...
+		LED_UCS1903,
+		LED_LPD1886_8bit,
+		LED_LPD1886_12bit
+	};
 
+	Controller mLEDController = LED_WS2811;
+	Channel mInputChannel = UNDEFINED_CHANNEL;
+
+	/// bits ber LED channel, either 8 or 10 at present
+	U8 BitSize() const;
```
Comment (Marcus10110) [inline source/AsyncRgbLedAnalyzerSettings.h]: Love these helper methods, it's a great pattern for these analyzers - historically we've done most of our switching in the analyzer worker thread. Also love the use of const.

File: source/AsyncRgbLedAnalyzerSettings.h [hunk 4]
```diff
@@ -10,18 +10,33 @@ class AsyncRgbLedAnalyzerSettings : public AnalyzerSettings
 	AsyncRgbLedAnalyzerSettings();
 	virtual ~AsyncRgbLedAnalyzerSettings();
 
-	virtual bool SetSettingsFromInterfaces();
+	bool SetSettingsFromInterfaces() override;
 	void UpdateInterfacesFromSettings();
-	virtual void LoadSettings( const char* settings );
-	virtual const char* SaveSettings();
+	void LoadSettings( const char* settings ) override;
+	const char* SaveSettings() override;
```
Comment (Marcus10110) [inline source/AsyncRgbLedAnalyzerSettings.h:23]: love the use of override

File: source/AsyncRgbLedAnalyzerSettings.h [hunk 5]
```diff
@@ -10,18 +10,33 @@ class AsyncRgbLedAnalyzerSettings : public AnalyzerSettings
 	AsyncRgbLedAnalyzerSettings();
 	virtual ~AsyncRgbLedAnalyzerSettings();
```
Comment (Marcus10110) [inline source/AsyncRgbLedAnalyzerSettings.h:11]: I'd like to start applying the rule of 5 to all of the user defined classes. We can start by deleting the copy & move special functions.
Quick note, you're probably aware, but if we default the destructor, we should put the "= default" in the cpp file, to ensure the destructor is compiled into the dll. Any defaulted function should be defaulted in the cpp file to ensure that it's built into the dll, and not inferred when used by the calling binary.

File: source/AsyncRgbLedAnalyzerSettings.h [hunk 6]
```diff
@@ -10,18 +10,33 @@ class AsyncRgbLedAnalyzerSettings : public AnalyzerSettings
...
+		LED_WS2812B,
+		LED_WS2813,
+		LED_TM1809,
+		LED_TM1804,
+		LED_UCS1903,
+		LED_LPD1886_8bit,
+		LED_LPD1886_12bit
+	};
 
+	Controller mLEDController = LED_WS2811;
```
Comment (Marcus10110) [inline source/AsyncRgbLedAnalyzerSettings.h]: love initialization in the class definition, I'll check to make sure this works with all of our targets. more detail in the overall review notes.

File: source/AsyncRgbLedAnalyzerSettings.h [hunk 7]
```diff
@@ -10,18 +10,33 @@ class AsyncRgbLedAnalyzerSettings : public AnalyzerSettings
 	AsyncRgbLedAnalyzerSettings();
```
Comment (Marcus10110) [inline source/AsyncRgbLedAnalyzerSettings.h:10]: I'd like to add `noexcept` to all methods where possible. I'll check if we can support that. More details in the overall review notes.

File: source/AsyncRgbLedAnalyzerSettings.h [hunk 8]
```diff
@@ -10,18 +10,33 @@ class AsyncRgbLedAnalyzerSettings : public AnalyzerSettings
...
 	void UpdateInterfacesFromSettings();
-	virtual void LoadSettings( const char* settings );
-	virtual const char* SaveSettings();
+	void LoadSettings( const char* settings ) override;
+	const char* SaveSettings() override;
 
-	
-	Channel mInputChannel;
-	U32 mBitRate;
+	enum Controller {
```
Comment (Marcus10110) [inline source/AsyncRgbLedAnalyzerSettings.h]: We should assign any serialized enums specific integer values.
1) we'll be serializating these settings and then deserializing them on different operating systems.
2) in the future we'll add new LED modules as they are released. It's important that the integer mappings don't change when an older saved file is loaded into a newer version of the analyzer.
If we assign integer values now, I think it it's less likely a future developer might inadvertently insert a new enum value (such as a hypothetical `LED_WS2814`) in the middle of the list, causing the integer mappings to change.

**Review by Marcus10110 — CHANGES_REQUESTED** [review]
This looks great! I added a few comments, mostly about what I like, with a few changes requested.

I will be running some tests on our build system this week to support these comments.

Our current build configurations:
1. Windows 10, MSVC 2015
2. MacOS 10.10.5, XCode 7.
3. Ubuntu 14, GCC 4.8.4. 

I *think* we build with `-std=c++11` on MacOS and on Linux, but our cmake setup for our current application is pretty messy, so I'll need to verify this.

I'll double check that we can build the following features and that the application still works on our target OSes:

- class declaration member initialization
- use of override
- use of noexcept
- use of move semantics
- use of `= default` and `= delete` for special functions
- use of std::unique_ptr

My main to concerns is that A) XCode 7 supports this, and B) this doesn't bump the minimum glibc / libstdc++ requirements past the point of support for Ubuntu 12.04.2.

If we have problems with either of these, I'll see about updating our build system or dropping support for Ubuntu 14.04 though, so please do use these features without waiting for my follow up.

File: source/AsyncRgbLedAnalyzerSettings.h (hunk 2 above)
Comment (zakalawe) [reply source/AsyncRgbLedAnalyzerSettings.h]: Done, I use unique_ptr by default anyway, this was copy-paste that I missed.

File: source/AsyncRgbLedAnalyzerSettings.cpp (hunk 1 above)
Comment (zakalawe) [reply source/AsyncRgbLedAnalyzerSettings.cpp]: Okay, will make it a switch i think.

File: source/AsyncRgbLedAnalyzerSettings.h (hunk 5 above)
Comment (zakalawe) [reply source/AsyncRgbLedAnalyzerSettings.h:11]: I'm assuming the destructor will be on-inline (and virtual) to pin the class to the .cpp file. I haven't been deleting the copy/move operators for heap-allocated classes but I can do that here, no worries.

File: source/AsyncRgbLedAnalyzerSettings.h (hunk 7 above)
Comment (zakalawe) [reply source/AsyncRgbLedAnalyzerSettings.h:10]: Okay, will try it.

File: source/AsyncRgbLedAnalyzerSettings.h (hunk 8 above)
Comment (zakalawe) [reply source/AsyncRgbLedAnalyzerSettings.h]: Sorry, I normally do this, when hardware is involved, just forgot here, thanks for catching.

**Comment by zakalawe** [comment]
Updated this, please re-check

**Review by Marcus10110 — APPROVED** [review]

**Pull request closed without merging**